### PR TITLE
feat(binary uuids): implement binary(16) uuids for cash table

### PR DESCRIPTION
**Pull request by jniles**
This commit implements a proof-of-concept binary UUID storage scheme for
the cash module.  The integration tests have been updated accordingly.

**Overview**
The introduction of binary uuids means that a little more work must be
done in the controllers and sql to insert/retrieve data.  Some commands
in Node and SQL have been added to ease this process.  I'll document
these differences as `before` and `after` scenarios.

READ

Before:

``` sql
SELECT uuid FROM table;
```

After:

``` sql
SELECT BUID(uuid) AS uuid FROM table;
```

**Note**: it is important to alias the column with `AS` in order to
preserve the property name `uuid`.  The command `BUID()` has been added
to our MySQL database instance to convert binary uuids into hexadecimal
uuids.

WRITE

Before:

``` js
var insertSql = 'INSERT INTO table SET ?;';
var updateSql = 'UPDATE table SET ? WHERE uuid = ?;';

var data = {
  uuid : /** some uuid */,
  /* other properties ... */
};

db.exec(insertSql, [ data ]).then(/* ... */).catch(/* .. */);
db.exec(updateSql, [ data, req.params.uuid ]).then(/* ... */).catch(/* .. */);
```

After:

``` js
var insertSql = 'INSERT INTO table SET ?;';
var updateSql = 'UPDATE table SET ? WHERE uuid = ?;';

var data = {
  uuid : /** some uuid */,
  /* other properties ... */
};

// convert the hex uuid into a binary uuid
data.uuid = db.bid(data.uuid);

// insert with the new data
db.exec(insertSql, [ data ]).then(/* ... */).catch(/* .. */);
db.exec(updateSql, [ data, db.bid(req.params.uuid) ]).then(/* ... */).catch(/* .. */);
```

I'd be curious if we can come up with a better API for binary
conversions.  The current method is pragmatic, but possibly unclear.

First implementation of #64.

### CODE REVIEW

**Comment by jniles** [comment]
@IMA-WorldHealth/local-contributors , what do you think of this method for binary UUIDs?  If this is approved, we should quickly change ALL uuids to binary uuids, not just the cash table uuids.

File: server/controllers/finance/cash.js [hunk 1]
```diff
@@ -35,25 +35,29 @@ exports.debitNote = debitNote;
...
+    'SELECT BUID(cash_item.uuid) AS uuid, cash_item.amount, cash_item.invoice_uuid ' +
     'FROM cash_item WHERE cash_item.cash_uuid = ?;';
 
-  var cashDiscardRecordSql =
-    'SELECT cash_uuid FROM cash_discard WHERE cash_uuid = ?;';
+  const cashDiscardRecordSql =
+    'SELECT BUID(cash_uuid) AS uuid FROM cash_discard WHERE cash_uuid = ?;';
+
+  // parse the uuid into a buffer
+  const buffer = db.bid(id);
```
Comment (DedrickEnc) [inline server/controllers/finance/cash.js:58]: Why not to have the same name as function, so that we should change the method `db.bid(id)` to `db.buid(id)`.

File: server/controllers/finance/cash.js [hunk 2]
```diff
@@ -140,19 +137,20 @@ function detail(req, res, next) {
...
- * generating it if it is not present.
+ * Creates a cash payment against one or many previous sales or a cautionary
+ * payment.  If a UUID is not provided, one is automatically generated.
  */
 function create(req, res, next) {
   'use strict';
 
-  var data = req.body.payment;
+  // alias insertion data
+  const data = req.body.payment;
```
Comment (DedrickEnc) [inline server/controllers/finance/cash.js]: Your use of `const` is so abusive, even if a constant is a value which will not be changed during the execution, but logic needs that a constant must be known and must remain the same in every situation.
so for example the area of circle is `area = 3.14 * (R*R)` in this case `3.14` is constant because it is known and will not change no mater the situation but `R` is the variable, because it depend on the problem, it can be `1, 2, 1.6, or 30 ...`.

**Comment by DedrickEnc** [comment]
@jniles 
I seen the implementation and I am interested, so I want to know if  you noticed any differency in terms of performance.
I am pretty sure that, we should get a good performance when fetching data `SELECT` query, but there is some additional steps for `INSERT, UPDATE` because we are not able to generate `BUID` directly.

File: server/controllers/finance/cash.js (hunk 2 above)
Comment (jniles) [inline server/controllers/finance/cash.js]: `const` in ECMAScript 6 does not mean that the value stays constant.  It means that the binding (the MDN calls it [the reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const)) should not change.  For example, see this code below:

``` js
const o = { value :1 };
console.log(o);  // logs { value : 1 }

o.value = 2;
console.log(o) : // logs { value : 2 }

delete o;  // returns false
console.log(o); // logs { value : 2 }
```

Clearly, the object `o` is not constant.  In this case, to actually make the value constant, you would need to use [Object.freeze](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze) as shown below:

``` js
const o = Object.freeze({ value : 1 });
console.log(o);  // logs { value : 1}

o.value = 2;
console.log(o); // logs { value : 1 }

delete o;  // returns false
console.log(o); // logs { value : 1 }
```

See [here](https://mathiasbynens.be/notes/es6-const) for more information.  See [here](http://stackoverflow.com/questions/33295772/does-it-make-sense-to-use-const-for-all-variables-that-will-never-be-changed) for a discussion on when to use `const`.

**Comment by jniles** [comment]
@DedrickEnc , if there was a performance difference, I couldn't measure it.  Since we never write more than a few rows at a time to the database, my guess is that the no user will ever feel a difference -- the difference would be fractions of a millisecond.

On read, I noticed slight increase in speed finding random binary UUIDs in a table of 25,000 rows -- but these differences were not statistically significant (too small sample size, too long to run tests).

File: server/controllers/finance/cash.js (hunk 1 above)
Comment (jniles) [inline server/controllers/finance/cash.js:58]: One function takes HEX -- > BIN, the other takes BIN --> HEX.  I was concerned that people might mix the two up if they had the same name.

If you feel strongly that they should be identically named, I can alter the API.

File: server/controllers/finance/cash.js (hunk 2 above)
Comment (DedrickEnc) [inline server/controllers/finance/cash.js]: I followed the two links, and I noticed that : 
1. You are right `const` just means that the binding should not change.
2. Using const that way is a choice and not a rule, according to this portion of the article

> Everything I wrote here so far are facts. What follows is entirely subjective, but bear with me.
>    Given the above, const makes code easier to read:
>    within its scope, a const variable always   refers to the same object.
>    With let there is no such  guarantee. 
>   As a result, it makes sense to use let and const as   follows in your ES6 code:
>  Use const by default
>  only use let if rebinding is needed
>  (var shouldn’t be used in ES6)

Here is the comment of @Andre Behrens

> Andre Behrens wrote on 5th February 2016 at 14:55: (reply)
>    I would say if we don’t want to fool people into thinking const is for constants, then const was a poor syntactic choice.
>    On our team, we treat it as a constant by convention, and everyone knows what that means. 
>    Kind of like _quasiPrivate variables. For us, const means “I intend this to be constant”.
>    It would be nice for the language not to have features that require this much explanation.

See also the second answer of stackoverflow bellow.

> Blindman67 gives a good argument from a performance viewpoint. Since javascript/ecmascript compilers are changing/improving all the time, I would consider answering this question more from the perspective of what gives more readable/maintainable code.
> 
> Marking a variable 'const' can be useful at times, signifying that a value is immutable. This stands also for the case when the compiler would not enforce const check - like marking an object array const would still allow you to alter state of an object in that array. This of course boils down to coding style within the team, but I found most people go out on that a variable defined const, even if could be changed, is meant to be immutable.
> 
> The other hand of this argument is overusing const. Of course you can use const everywhere, it just makes the code somewhat harder to wade through, and harder to make changes to it. I've found that in practice, most maintenance work of such code begins with removing most of the 'const' keywords, so that the necessary extra changes could be done. So while having lots of constants might look like great code at first, in the longer run I think it adds close to nothing to maintainability, on the contrary, it makes maintenance harder.
> 
> tl;dr Use const where you think it adds to the readability of your code.

Comment (jniles) [inline server/controllers/finance/cash.js]: So, what is your conclusion?

Comment (DedrickEnc) [inline server/controllers/finance/cash.js]: Let keep our philosophy, let use `const` if we are expecting the value to not change, otherwise let use `var`.

Comment (jniles) [inline server/controllers/finance/cash.js]: Would you be okay if I used `let` instead?  I think ECMAScript is going to try and leave `var` behind.

**Comment by jniles** [comment]
@dedrickenc, can I get a review?

File: server/controllers/finance/cash.js (hunk 2 above)
Comment (DedrickEnc) [inline server/controllers/finance/cash.js]: `var` is more expressive

**Comment by jniles** [comment]
@dedrickenc, is preferring `let` over `var` really going to block this PR?  Or are there other problems with creating binary uuids?

**Comment by DedrickEnc** [comment]
@jniles 
It is more than that, you are about to introduce a new way of declaring variable and constant.
So it is not a small thing,  a team must have a philosophy for coding.
The **binary uuid is ok with me**, just use `var` and I will merge the PR.
We will discuss about using `var`, `let`, and  `const` when every one will be at kinshasa at april if that is okay with you.

**Comment by jniles** [comment]
Okay, we can discuss when everyone's back.

I will not change this because:
1. It is valid JavaScript and supported in all our targeted versions.
2. It is more precise to use a variable that is block-scoped.  `var` is hoisted to the top of the function declaration, which is one of the strange design decisions in JavaScript.  So, I've used both `const` for variables that are immutable and `let` for variables that are mutable.  If it is of interest, [here is a comparison of let versus var](https://davidwalsh.name/for-and-against-let).
3. @mbayopanda asked if we could use these new JavaScript features.  We agreed as a team that we could use these.  Now you have changed your mind, and it is unclear why.  Just because it is new, does not mean it is bad.

I am deeply concerned in the number of pull requests that are blocked from comments not related to their:
1. Features
2. Performance
3. Error handling
4. Tests

These pull requests seem to be blocked for silly reasons, usually their choice of names or something similar.  For this software to grow, we should learn how to prioritize the functionality over the code style.  Particularly when we have not had discussions to say that one style is better than another.

I will reopen this pull request when we have a chance to discuss this.

**Comment by sfount** [comment]
As we are running out of time I am making the decision to open this pull request and I will evaluate the code. As we are using git to version control our codebase this can be undone if necessary once the team is together.

**Comment by jniles** [comment]
@sfount, updated to latest master.

File: server/controllers/finance/cash.js [hunk 3]
```diff
@@ -35,25 +35,29 @@ exports.debitNote = debitNote;
 
 // looks up a single cash record and associated cash_items
 // sets the "canceled" flag if a cash_discard record exists.
-function lookupCashRecord(uuid, codes) {
+function lookupCashRecord(id, codes) {
```
Comment (sfount) [inline server/controllers/finance/cash.js:38]: Just a note that this should be updated given https://github.com/IMA-WorldHealth/bhima-2.X/issues/182. This is not directly related to the pull request and will not block it. 

This should be addressed in the Cash 2.x Review.

**Comment by sfount** [comment]
To conclude the discussion in this thread before I close this; the team will discuss the use of newly introduced JavaScript features when they return from Tshikaji. Right now we are not in a position (both with geographical differences and time constraints) to have everything fully worked out before integration. 

As everything in this pull request is supported in the controlled nodejs environment that we support I am confident that this code will work and pass all of the tests. If the team overrules this from a maintainability standpoint the code can be updated.
